### ISSUE TITLE
Kernel-backed file viewer plugin

### ISSUE THREAD

**Issue by dopplershift** [jupyterlab/jupyterlab#448]
(Apologies if this is covered anywhere.)

It looks like the current file viewer plugins are all js/typescript based (and admittedly cool). What would be really awesome for my use is having the ability to spin up/talk to a kernel to do the display. I'm specifically thinking of being able to view a netCDF4 file, where I really want no parts of coding in js.

Not sure if this is in scope or asking for the moon.

**Comment by jasongrout** [comment]
We essentially do something like this for ipython widgets - since it's an output in a notebook, we already have access to a kernel. I bet the astronomy people would also like to have something like this for exploring FITS images that are too big for javascript to handle.

So, like widgets, we could write a python piece and a typescript piece, and they could communicate via comms in the context of an already running kernel.

**Comment by dopplershift** [comment]
Right, I figure the communications isn't the hard part at this point, just having the jupyterlab plugin infrastructure be able to give us this "for free".

So many ideas...I need to learn how to sleep less.

**Comment by jasongrout** [comment]
Alternatively (probably more tricky), a server extension could be installed in Jupyter which would provide an api for communicating data to a typescript frontend for the data. When a netCDF4 file is clicked, the launched typescript component pings the server api to start one of these data file handlers. Then it starts communicating with the data file handler directly, fetching and displaying whatever it wants.

I bet the HDF5 people, and the astronomy FITS-using people, would love to have this too.

**Comment by jasongrout** [comment]
We'd need to build in the option to define more ways of requesting a file in the ContentsManager. Right now, I don't know if there is a way to say 'open the file *without* trying to transfer the data'.

**Comment by dopplershift** [comment]
Ah, I see. I sometimes forget to think of this as a full server-client model, where the server really could/should/is not where I'm running javascript.

**Comment by jasongrout** [comment]
I'm not too familiar with netCDF4 files, but with HDF5 files, I could imagine a server component would initially just transmit the hierarchal structure of the file, which you could then explore. Clicking on a particular data set would transfer just a window of the data to a grid widget, which you would update as the user scrolled in the dataset. Standard virtual grid stuff for big files, I suppose.

**Comment by jasongrout** [comment]
yeah, you definitely want to have a server component communicating the viewing data to typescript - I don't want to write a netCDF4 viewer in javascript...

**Comment by jasongrout** [comment]
hehe, someone already has, apparently: https://www.npmjs.com/package/netcdf4

**Comment by danielballan** [comment]
@dopplershift This feature is near the top of my own wishlist. BNL has a lot of very large HDF5 files (in various layouts) and loading all the data into the browser is not an option.

**Comment by johanfforsberg** [comment]
Being able to browse HDF5 files this way would be great! I agree that loading the full files into the browser in general would not be feasible, as in our case they might be arbitrarily large.

Don't know if it's relevant, but the HDF group has a HTTP interface to HDF5 (in python): https://www.hdfgroup.org/projects/hdfserver/

**Comment by danielballan** [comment]
@johanfforsberg It would be nice to reuse some part of that if possible. Yesterday, @jasongrout and I discussed the benefits of reusing the kernel architecture, at least in an initial implementation. That is, making a special kernel for sending pieces of HDF5 files back and forth.

**Comment by johanfforsberg** [comment]
@danielballan Interesting, so you mean that there would be a "hdf5" kernel that you would run for each hdf5 file you want to access? I'm not familiar enough with the kernel architecture to have a feeling for how that would work in practice but it sounds like a cool idea.

In many ways, a HDF5 file is more like a filesystem than a file, so maybe it makes more sense to handle it specifically. Perhaps that would also make sense for other similar things, like netCDF, sqlite...

BTW, this seems to be what h5serv is using for serialising HDF5 over JSON: http://hdf5-json.readthedocs.io/en/latest/

**Comment by jasongrout** [comment]
> so you mean that there would be a "hdf5" kernel that you would run for each hdf5 file you want to access?

Yes, that was the idea on the table. The issue here is that we may want two-way communication with the python side of an hdf5 file, in which case taking advantage of the existing architecture for kernel websockets may make sense.

On the other hand, if we can fetch all that we need via a rest api, perhaps it is enough to extend the contents manager with a plugin that can fetch the appropriate data.

Probably the first thing to do is to highlight important use cases for dealing with hdf5 (or similar) files in JupyterLab. For example, I think it makes sense to be able to:

1. browse the hierarchal data structure as a tree
2. fetch a small slice of data from a dataset (so we can have a grid that updates lazily as the user scrolls through it)
3. Be able to update single values in a data set nested in the file.
4. Do all this without transferring the entire file across the network, or writing a parser in Javascript.

**Comment by johanfforsberg** [comment]
Off the top of my head, that list should cover most use cases. 

For my particular use case, no. 3 is the least important. This is for experimental data that should only be writable by the acquisition process, and read-only for everyone else. I think just being able to browse structure and read metadata would be very useful as a start. 

Also, writing to a HDF5 file is not totally uncomplicated. Normally it's not possible to have different clients both reading and writing the same file. There's some support called "SWMR" (single writer, multiple reader) but it's a very new feature with some limitations. In fact I think this is one reason that the HDF group is developing things like h5serv, where all the access to a file would go through the same process, circumventing the issue. 

I suppose "real" analysis of the data would happen in a notebook, so maybe the viewer should focus on quick navigation and "at a glance" information. For large datasets, a table view is probably kind of useless anyway. Does it make sense to somehow make a connection between the viewer and the notebook, so that e.g. one could drag'n'drop a dataset from the viewer and magically get a reference of some kind in the notebook? :)

**Comment by jasongrout** [comment]
All good points about being read-only (at least an initial version). The tricky part with the drag/drop references is probably writing the python/julia/r/whatever code in the frontend to access the object.

**Comment by jakirkham** [comment]
Was just thinking about this problem today. Though expect my use cases are a bit different (though not totally). Given it's been almost been a year (not quite) since the bulk of the discussion occurred and some of the issues associated may have been tackled since, would be curious to know where we stand now in terms of working with this "larger than JavaScript can handle" data.

**Comment by jasongrout** [comment]
I don't know of anyone working on this. I'm happy to contribute to the discussion if someone did pick it up.

**Comment by jasongrout** [comment]
(sorry, hit the wrong button)

**Comment by jakirkham** [comment]
Saw this young HDF5 viewer widget, which might be interesting. It only shows metadata about the content. Though it certainly could be built off of.

ref: https://github.com/m-rossi/hdf5widget

**Comment by dnelson86** [comment]
Hi,

Found this after starting to work with JupyterLab and wondering about a HDF5 viewer. My use cases are similar as above, large (1 GB to multi-TB) HDF5 files, for which some basic introspection would be nice (read-only definitely the place to start). These cannot be transferred, so I agree with the idea to use the existing bidirectional kernel communication infrastructure as an option for plugins.

In particular, as noted above, an 'Open' file command should first check with the relevant viewer if this means (a) transfer raw to client, or (b) process on the server and send something else. In the case of (b), the viewer would need a kernel, spawning as necessary, and then the two components (server and client) of the viewer would need to be able to communicate.

I don't think restricting the scope to JSON (or REST API) type communication is a good idea, for e.g. HDF5 subsets binary data transfer to the browser might be required.

Just some thoughts, hope this continues towards an implementation.

**Comment by dnelson86** [comment]
See also phosphorjs/phosphor/pull/307

**Comment by danielballan** [comment]
> I don't think restricting the scope to JSON (or REST API) type communication is a good idea, for e.g. HDF5 subsets binary data transfer to the browser might be required.

It's a fair point. As you probably discovered in my phosphorjs PR, that one makes a separate REST call for each fetched block of data. One could imagine using websockets instead. As Jason has pointed out before, Jupyter Kernel messages are quite a bit heavier than the minimal message one could imagine using for this purpose. For best performance, something like h5serv + custom websocket messages, are the likely winner.

**Comment by jakirkham** [comment]
Has anyone here seen [h5serv]( https://github.com/HDFGroup/h5serv )?

**Comment by danielballan** [comment]
Yeah, it looks promising to me. It's what phosphorjs/phosphor#307 works on.

**Comment by jakirkham** [comment]
Ah, sorry, hadn't looked at the PR yet. Clearly this is already obvious to you. :)

**Comment by danielballan** [comment]
Haha, no worries @jakirkham. :-D

**Comment by jasongrout** [comment]
The other problem with using a kernel is that the kernels a system provides may not be anywhere close to the filesystem the notebook is providing. The filesystem is provided at the notebook server level, not the kernel level, so I think it makes sense to make the thing interpreting files provided at the notebook server level. In other words, I think rather than having a special kernel that provides file streaming, it's much better to build a notebook server extension that provides it.
